### DISCUSSION TITLE
Uploading images in Powershell

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [buttondown/discussions#26] [Show & tell]
I've started writing my emails in Visual Studio Code because I can write in Markdown and use my local tools to assist with the content. When I finish writing, I copy and paste the Markdown text into the web editor. This worked fine until I needed to insert a screenshot. My solution is this PowerShell function. PowerShell is my area of expertise so I thought I'd share it here.

```powershell
# New-ButtondownImage.ps1
Function New-ButtonDownImage {
    [CmdletBinding(SupportsShouldProcess)]
    [OutputType([string])]
    [Alias('bdi')]
    Param(
        [Parameter(Position = 0, Mandatory)]
        [ValidateNotNullOrEmpty()]
        [ValidateScript({ (Test-Path $_) -AND $_.extension -match 'png|gif|jpg' })]
        [ArgumentCompleter({
                [OutputType([System.Management.Automation.CompletionResult])]
                param(
                    [string] $CommandName,
                    [string] $ParameterName,
                    [string] $WordToComplete,
                    [System.Management.Automation.Language.CommandAst] $CommandAst,
                    [System.Collections.IDictionary] $FakeBoundParameters
                )

                (Get-ChildItem -Path . -File).Where({ $_.extension -match 'png|gif|jpg' }) |
                ForEach-Object {
                    [System.Management.Automation.CompletionResult]::new($_, $_, 'ParameterValue', $_.FullName)
                }
            })]
        [System.IO.FileInfo]$FilePath,
        [Parameter(Position = 1, Mandatory)]
        [string]$AltText,

        [Parameter(Position = 2)]
        [string]$Caption,

        [Parameter(Mandatory)]
        [ValidateNotNullOrEmpty()]
        [ValidateLength(36, 36)]
        [string]$APIKey
    )

    $imagePath = Convert-Path $FilePath

    if (-Not $PSBoundParameters.ContainsKey('Caption') -AND $global:bdImageCount -ge 0) {
        $caption = "figure $global:bdImageCount"
    }

    `Write-Host "Uploading $imagePath ($altText) [$caption]" -ForegroundColor Yellow

    If ($PSCmdlet.ShouldProcess("$imagePath ($altText)" )) {
        $r = curl -s -X POST -H "Authorization: Token $APIKey" -H 'Content-Type: multipart/form-data' -F "image=@$ImagePath" 'https://api.buttondown.email/v1/images'
        $obj = $r | ConvertFrom-Json

        if ($obj.Image) {
            $src = $obj.image
            $link = '<figure><img src="{0}?w=960&amp;fit=max" alt="{1}" draggable="false" contenteditable="false"><figcaption>{2}</figcaption></figure>' -f $src, $AltText, $caption
            $link | Set-Clipboard
            $link | Tee-Object -Variable global:img
            if (-Not $PSBoundParameters.ContainsKey('Caption')) {
                #only increment the counter if we are not using a custom caption
                $global:bdImageCount++
            }
        }
        else {
            Write-Warning 'Failed to upload image'
        }
    }
}

#set a global variable to track the number of images uploaded for Figure captions
$global:bdImageCount = 1
```

If you are new to PowerShell, you need to dot-source the .ps1 script file to load the function into your PowerShell session. You also need to pass your API key. I store mine in a Microsoft secrets vault. I have additional code in the script file to get the key and set it as the default parameter value for this command.

```powershell
$bdAPI = Get-Secret -Name buttondown-api -Vault $myVault -AsPlainText
$PSDefaultParameterValues['New-ButtonDownImage:APIKey'] = $bdAPI
```

This way, I don't have to hardcode the API key in the script, which is bad, and I don't have to type it out. But you can use whatever technique you want to pass the API key to the function.

To use the function, I need to specify the path to the image file. The function will autocomplete valid image files in the current directory. I use the function's alias to save a little typing. I can type `bdi <ctrl-space>` to have PowerShell show me the list of images.

![image](https://github.com/buttondown/discussions/assets/12089920/95802f06-0d95-41e9-bac4-01f894b9a15c)

I select the one I want and press `Enter`.  I want my images to have ALT text so I can either use the `-AltText` parameter or press `Enter` and PowerShell will prompt me for the text. The command has a parameter to also let you insert a caption. By default, I autogenerate figure captions, i.e., `figure 1`.

After uploading the image, the function generates the text I need for the Markdown document. It writes it to the pipeline and copies it to the clipboard. 

```html
<figure><img src="https://assets.buttondown.email/images/201de8fe-36c4-4ad9-adac-c281d7058f7f.png?w=960&amp;fit=max" alt="System.Windows.Window type" draggable="false" contenteditable="false"><figcaption>figure 1</figcaption></figure>
```
All I need to do is paste the image code into my document. Now, I can paste the complete document into the web editor and preview it.

I know I could use the API to draft the email, and I might look at that in the future.

**Comment by jdhitsolutions** [comment]
Savvy PowerShell users will note that the code uses `curl`. Normally, you would use `Invoke-Restmethod`.  However, there is something weird about how the cmdlet implements `POST` with multi-part data. Rather than spend any more time figuring out why, it was quicker to invoke `curl`. I'm using PowerShell to build the `curl` command.

**Comment by jdhitsolutions** [comment]
If you add a `Show and Tell` discussion category, I can move this topic there. I'm happy to share my PowerShell-based management solutions with others.

**Reply by jmduke** [reply]
Aaaand done!

**Comment by jdhitsolutions** [comment]
A quick tip if you are using VSCode to generate your markdown and then want to paste it into the Buttondown web UI. Don't copy and paste from the VS Code window. You'll end up with extra hidden code.  You can either copy and paste into Notepad and then re-copy and paste in the UI. If you have a PowerShell prompt handy you can also run `Get-Content fabulous.md | Set-ClipBoard`. Don't forget to set the web editor to use Markdown.
